File: backend/services/workbench/managed-tool-policy.js
const MUTATING_NAME_PATTERN = /write|edit|create|move|rename|delete|remove|install|import|save|set|update|add|forget|remember|bash|command|run|spawn|launch|click|type|focus|clipboard_set|patch|observe|link|index|store|learn/i;
const SAFE_NAME_PATTERN = /read|list|search|fetch|get|describe|diagnose|status|recall|review|scan|preview|find|entities/i;

const SAFE_AUGUST_TOOLS = new Set([
    'august__read_file',
    'august__search_past_conversations',
    'august__recall',
    'august__list_facts',
    'august__review_learned_guidelines',
    'august__graph_recall',
    'august__graph_entities',
    'august__list_agent_jobs',
    'august__get_agent_job',
    'august__find_skill_sources',
    'august__preview_skill_import',
    'august__load_skill',
    'august__scan_brain'
]);

function isSupermemoryRead(args = {}) {
    return ['search', 'list'].includes(String(args.action || '').toLowerCase());
}

function isManagedToolParallelSafe(toolName, args = {}) {
    const name = String(toolName || '');
    if (!name) return false;
    if (name === 'WebSearch' || name === 'WebFetch' || name === 'web_search' || name === 'web_fetch') return true;
    if (name === 'mcp__workspace__web_search' || name === 'mcp__workspace__web_fetch') return true;
    if (name === 'august__supermemory') return isSupermemoryRead(args);
    if (SAFE_AUGUST_TOOLS.has(name)) return true;
    if (name.startsWith('august__')) return false;
    if (name.startsWith('computer_')) return /screenshot|position|screen_size|list_windows|clipboard_get/i.test(name);
    if (name.startsWith('mcp__') || name.startsWith('cowork_') || name.startsWith('mcp__cowork__')) {
        if (MUTATING_NAME_PATTERN.test(name)) return false;
        return SAFE_NAME_PATTERN.test(name);
    }
    if (MUTATING_NAME_PATTERN.test(name)) return false;
    return SAFE_NAME_PATTERN.test(name);
}

function parseOpenAiToolArgs(toolCall) {
    try {
        return JSON.parse(toolCall?.function?.arguments || '{}');
    } catch (_err) {
        return {};
    }
}

function isOpenAiToolCallParallelSafe(toolCall) {
    return isManagedToolParallelSafe(toolCall?.function?.name, parseOpenAiToolArgs(toolCall));
}

function isAnthropicToolUseParallelSafe(toolUse) {
    return isManagedToolParallelSafe(toolUse?.name, toolUse?.input || {});
}

module.exports = {
    isAnthropicToolUseParallelSafe,
    isManagedToolParallelSafe,
    isOpenAiToolCallParallelSafe,
    parseOpenAiToolArgs
};

File: backend-py/app/services/workbench/managed_tool_policy.py
"""
Managed tool policy — parallel safety checks for tool execution.

Port of backend/services/workbench/managed-tool-policy.js (63 lines).

Determines which tools are safe to run in parallel vs. mutating
(read-only vs. write).
"""

from __future__ import annotations
import json
import re
from app.typeAliases import JsonValue

MUTATING_NAME_PATTERN = re.compile(
    '^(write|edit|create|delete|install|run|bash|launch|click|type|focus|set|add|remove|rename|copy|move|mkdir|touch|chmod|kill|uninstall|stop|restart|upload|download|patch|apply|commit|push|merge|deploy|start|reboot|shutdown|format|mount|unmount|browser_navigate|browser_click|browser_type|browser_snapshot)',
    re.IGNORECASE,
)
SAFE_NAME_PATTERN = re.compile(
    '^(read|list|search|fetch|get|describe|diagnose|status|recall|view|find|show|check|inspect|lookup|resolve|ping|health|info|memory_search|fact_search|context_read|list_skills)',
    re.IGNORECASE,
)


def isManagedToolParallelSafe(toolName: str, args: dict[str, object] | None = None) -> bool:
    """Check if a managed tool is safe to run in parallel.

    Returns True for:
    - WebSearch/WebFetch (and workspace MCP variants)
    - August supermemory reads (action: 'search' or 'list')
    - Safe August tools (read_file, search, recall, list, review, etc.)
    - Read-only computer tools (screenshot, position, screen size, etc.)
    - MCP/Cowork tools that match SAFE_NAME_PATTERN and not MUTATING_NAME_PATTERN
    """
    if not toolName:
        return False
    if toolName in (
        'WebSearch',
        'WebFetch',
        'web_search',
        'web_fetch',
        'mcp__workspace__web_search',
        'mcp__workspace__web_fetch',
    ):
        return True
    if MUTATING_NAME_PATTERN.match(toolName):
        return False
    if SAFE_NAME_PATTERN.match(toolName):
        return True
    return False


def isOpenaiToolCallParallelSafe(toolCall: dict[str, object]) -> bool:
    """Check if an OpenAI-format tool call is parallel-safe."""
    func = toolCall.get('function', {})
    assert isinstance(func, dict)
    name = func.get('name', '')
    assert isinstance(name, str)
    argsStr = func.get('arguments', '{}')
    try:
        args = json.loads(argsStr) if isinstance(argsStr, str) else argsStr
    except (json.JSONDecodeError, TypeError):
        args = {}
    assert isinstance(args, dict)
    return isManagedToolParallelSafe(name, args)


def isAnthropicToolUseParallelSafe(toolUse: dict[str, object]) -> bool:
    """Check if an Anthropic-format tool use is parallel-safe."""
    name = toolUse.get('name', '')
    assert isinstance(name, str)
    args = toolUse.get('input', {})
    assert isinstance(args, dict)
    return isManagedToolParallelSafe(name, args)


def parseOpenaiToolArgs(toolCall: dict[str, object]) -> dict[str, object]:
    """Parse the function.arguments JSON string from an OpenAI tool call."""
    func = toolCall.get('function', {})
    if not isinstance(func, dict):
        return {}
    argsStr = func.get('arguments', '{}')
    if isinstance(argsStr, str):
        try:
            result = json.loads(argsStr)
            assert isinstance(result, dict)
            return result
        except (json.JSONDecodeError, TypeError):
            return {}
    return argsStr if isinstance(argsStr, dict) else {}
